onnxruntime/core/codegen/mti/math/halide_ops.h

```diff
@@ -0,0 +1,47 @@
+// Halide Copyright info:
+/*
+Copyright (c) 2012-2018 MIT CSAIL, Google Inc., and other contributors
+
+Developed by:
+
+  The Halide team
+  http://halide-lang.org
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of
+this software and associated documentation files (the "Software"), to deal in
+the Software without restriction, including without limitation the rights to
+use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+of the Software, and to permit persons to whom the Software is furnished to do
+so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// This file contains some ops that were copied from Halide (with small modifications).
+
+#pragma once
+#include <topi/elemwise.h>
+
+namespace onnxruntime {
+namespace tvm_codegen {
+
+tvm::Expr halideir_exp(const tvm::Expr& x_full);
+
+tvm::Expr halideir_log(const tvm::Expr& x_full);
+
+tvm::Expr fast_log(const tvm::Expr& x);
```

Comment:
I beleive those are also for CPU only, right?

Reply:
For non-hw accelerated only.  It is a polynomial special function calculation.
If there are hw-accelerated instruction, like FPGA usually building its own lookup table for SF.
Then, it is not applicable.

Reply:
Moved out

---
In reply to: [294524957](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294524957) [](ancestors = 294524957)